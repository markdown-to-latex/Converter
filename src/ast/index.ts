export * as node from './node';
export * as parsing from './parsing';
export * from './nodes';
