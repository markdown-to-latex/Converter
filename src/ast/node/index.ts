export * from './extension';
export * from './function';
export * from './position';
export * from './struct';
