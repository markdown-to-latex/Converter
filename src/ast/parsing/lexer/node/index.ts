export * as blockquote from './blockquote';
export * as breaks from './breaks';
export * as code from './code';
export * as codeSpan from './codeSpan';
export * as comment from './comment';
export * as del from './del';
export * as em from './em';
export * as formulaOrLatex from './formulaOrLatex';
export * as formulaSpan from './formulaSpan';
export * as heading from './heading';
export * as hr from './hr';
export * as image from './image';
export * as latexSpan from './latexSpan';
export * as link from './link';
export * as list from './list';
export * as macros from './macros';
export * as space from './space';
export * as strong from './strong';
export * as table from './table';
export * as underline from './underline';
