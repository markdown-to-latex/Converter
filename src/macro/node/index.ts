import table from './processing/table';
import code from './processing/code';
import { NodeProcessed, ProcessingInfo } from './struct';
import {
    Node,
    NodeEParentData,
} from '../../ast/node';
import { ContextE } from '../../context';
import picture from './processing/picture';

const ALL_PROCESSING: ProcessingInfo<Node>[] = [
    picture as ProcessingInfo<Node>,
    table as ProcessingInfo<Node>,
    code as ProcessingInfo<Node>,
];

export function processNode(
    ctx: ContextE,
    data: NodeEParentData,
): NodeProcessed[] | null {
    const node = data.node.n;
    const processing = ALL_PROCESSING.find(d => d.type === node.type);

    return !processing
        ? null
        : processing.callback(ctx, {
              node: data.node,
              index: data.index,
              container: data.container,
          });
}
