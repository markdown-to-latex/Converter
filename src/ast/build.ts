import { FileNode, getNodeAllChildren, Node, NodeType } from './nodes';
import { TokenListContainer } from '../lexer/tokens';
import { astProcessTokenList } from './visitors';

function updateASTParents(ast: Node) {
    interface NodeInfo {
        parent: Node | null;
        node: Node;
    }

    const nodeQueue: NodeInfo[] = [];
    nodeQueue.push({
        parent: null,
        node: ast,
    });

    // In-depth AST traverse
    while (nodeQueue.length !== 0) {
        const info = nodeQueue.pop()!;

        // Update parent
        info.node.parent = info.parent;

        nodeQueue.push(
            ...getNodeAllChildren(
                info.node as Parameters<typeof getNodeAllChildren>[0],
            ).map(node => ({
                parent: info.node,
                node: node,
            })),
        );
    }
}

export interface MarkdownASTData {
    filepath?: string;
}

export function buildMarkdownAST(
    tokens: Readonly<TokenListContainer>,
    options: Readonly<MarkdownASTData>,
): FileNode {
    const ast: FileNode = {
        type: NodeType.File,
        parent: null,
        children: astProcessTokenList(tokens),
        path: options.filepath ?? '__not-specified__',
    };
    updateASTParents(ast);

    return ast;
}
