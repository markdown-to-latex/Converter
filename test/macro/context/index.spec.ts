import {
    ContextApplicationContentInfo,
    contextDataGenericEmpty,
    ContextE,
    ContextPictureInfo,
    ContextReferenceContentInfo,
    ContextTableInfo,
} from '../../../src/macro/context';
import { Node, NodeType, TextNode } from '../../../src/ast/node';

function createSimpleTextNode(text: string): TextNode {
    return {
        type: NodeType.Text,
        parent: null,
        pos: {
            start: 0,
            end: text.length,
        },
        text,
    };
}

function createSimpleContext(): ContextE {
    return new ContextE({
        temp: {
            node: {
                type: NodeType.File,
                pos: {
                    start: 0,
                    end: 1,
                },
                parent: null,
            },
            application: null,
            reference: null,
            table: null,
        },
        data: {
            application: contextDataGenericEmpty(),
            reference: contextDataGenericEmpty(),
            picture: contextDataGenericEmpty(),
            table: contextDataGenericEmpty(),
            formula: contextDataGenericEmpty(),
        },
        diagnostic: [],
        // writeDiagnosticList: () => {},
        // writeFile: () => {},
    });
}

describe('context pictures', () => {
    test('simple', () => {
        const context = createSimpleContext();

        context.createPictureLabelData({
            label: createSimpleTextNode('pic-3'),
        } as any as ContextPictureInfo);
        context.getOrCreatePictureLabelIndex(createSimpleTextNode('pic-1'));
        context.getOrCreatePictureLabelIndex(createSimpleTextNode('pic-2'));
        context.getOrCreatePictureLabelIndex(createSimpleTextNode('pic-3'));
        context.createPictureLabelData({
            label: createSimpleTextNode('pic-2'),
        } as any as ContextPictureInfo);
        context.createPictureLabelData({
            label: createSimpleTextNode('pic-1'),
        } as any as ContextPictureInfo);

        context.diagnoseAll();

        expect(context.c.diagnostic).toHaveLength(0);
        expect(context.c.data.picture).toMatchSnapshot();
    });

    test('create same key twice', () => {
        const context = createSimpleContext();

        context.createPictureLabelData({
            label: createSimpleTextNode('pic-1'),
        } as any as ContextPictureInfo);
        context.createPictureLabelData({
            label: createSimpleTextNode('pic-1'),
        } as any as ContextPictureInfo);

        context.diagnoseAll();

        expect(context.c.diagnostic).toHaveLength(1);
        expect(context.c.diagnostic).toMatchSnapshot();
        expect(context.c.data.picture).toMatchSnapshot();
    });

    test('create undefined keys', () => {
        const context = createSimpleContext();

        context.getOrCreatePictureLabelIndex(createSimpleTextNode('pic-1'));
        context.getOrCreatePictureLabelIndex(createSimpleTextNode('pic-3'));

        context.diagnoseAll();

        expect(context.c.diagnostic).toHaveLength(2);
        expect(context.c.diagnostic).toMatchSnapshot();
        expect(context.c.data.picture).toMatchSnapshot();
    });

    test('create unused keys', () => {
        const context = createSimpleContext();

        context.createPictureLabelData({
            label: createSimpleTextNode('pic-1'),
        } as any as ContextPictureInfo);
        context.createPictureLabelData({
            label: createSimpleTextNode('pic-2'),
        } as any as ContextPictureInfo);

        context.diagnoseAll();

        expect(context.c.diagnostic).toHaveLength(2);
        expect(context.c.diagnostic).toMatchSnapshot();
        expect(context.c.data.picture).toMatchSnapshot();
    });
});

describe('context tables', () => {
    test('simple', () => {
        const context = createSimpleContext();

        context.createTableLabelData({
            label: createSimpleTextNode('table-3'),
        } as any as ContextTableInfo);
        context.getOrCreateTableLabelIndex(createSimpleTextNode('table-1'));
        context.getOrCreateTableLabelIndex(createSimpleTextNode('table-2'));
        context.getOrCreateTableLabelIndex(createSimpleTextNode('table-3'));
        context.createTableLabelData({
            label: createSimpleTextNode('table-2'),
        } as any as ContextTableInfo);
        context.createTableLabelData({
            label: createSimpleTextNode('table-1'),
        } as any as ContextTableInfo);

        context.diagnoseAll();

        expect(context.c.diagnostic).toHaveLength(0);
        expect(context.c.data.table).toMatchSnapshot();
    });

    test('create same key twice', () => {
        const context = createSimpleContext();

        context.createTableLabelData({
            label: createSimpleTextNode('table-1'),
        } as any as ContextTableInfo);
        context.createTableLabelData({
            label: createSimpleTextNode('table-1'),
        } as any as ContextTableInfo);

        context.diagnoseAll();

        expect(context.c.diagnostic).toHaveLength(1);
        expect(context.c.diagnostic).toMatchSnapshot();
        expect(context.c.data.table).toMatchSnapshot();
    });

    test('create undefined keys', () => {
        const context = createSimpleContext();

        context.getOrCreateTableLabelIndex(createSimpleTextNode('table-1'));
        context.getOrCreateTableLabelIndex(createSimpleTextNode('table-3'));

        context.diagnoseAll();

        expect(context.c.diagnostic).toHaveLength(2);
        expect(context.c.diagnostic).toMatchSnapshot();
        expect(context.c.data.table).toMatchSnapshot();
    });

    test('create unused keys', () => {
        const context = createSimpleContext();

        context.createTableLabelData({
            label: createSimpleTextNode('table-1'),
        } as any as ContextTableInfo);
        context.createTableLabelData({
            label: createSimpleTextNode('table-2'),
        } as any as ContextTableInfo);

        context.diagnoseAll();

        expect(context.c.diagnostic).toHaveLength(2);
        expect(context.c.diagnostic).toMatchSnapshot();
        expect(context.c.data.table).toMatchSnapshot();
    });
});

describe('context application', () => {
    test('simple', () => {
        const context = createSimpleContext();

        context.createApplication({
            label: createSimpleTextNode('app-1'),
        } as any as ContextApplicationContentInfo);
        context.createApplication({
            label: createSimpleTextNode('app-2'),
        } as any as ContextApplicationContentInfo);
        expect(
            context.getApplicationLabelIndex(createSimpleTextNode('app-2')),
        ).toEqual(0);
        expect(
            context.getApplicationLabelIndex(createSimpleTextNode('app-1')),
        ).toEqual(1);

        context.diagnoseAll();

        expect(context.c.diagnostic).toHaveLength(0);
        expect(context.c.data.application).toMatchSnapshot();
    });

    test('create same key twice', () => {
        const context = createSimpleContext();

        context.createApplication({
            label: createSimpleTextNode('app-1'),
        } as any as ContextApplicationContentInfo);
        context.createApplication({
            label: createSimpleTextNode('app-1'),
        } as any as ContextApplicationContentInfo);

        context.diagnoseAll();

        expect(context.c.diagnostic).toHaveLength(2);
        expect(context.c.diagnostic).toMatchSnapshot();
        expect(context.c.data.application).toMatchSnapshot();
    });

    test('create undefined keys', () => {
        const context = createSimpleContext();

        context.getApplicationLabelIndex(createSimpleTextNode('app-1'));
        context.getApplicationLabelIndex(createSimpleTextNode('app-2'));

        context.diagnoseAll();

        expect(context.c.diagnostic).toHaveLength(2);
        expect(context.c.diagnostic).toMatchSnapshot();
        expect(context.c.data.application).toMatchSnapshot();
    });

    test('create unused keys', () => {
        const context = createSimpleContext();

        context.createApplication({
            label: createSimpleTextNode('app-1'),
        } as any as ContextApplicationContentInfo);
        context.createApplication({
            label: createSimpleTextNode('app-2'),
        } as any as ContextApplicationContentInfo);

        context.diagnoseAll();

        expect(context.c.diagnostic).toHaveLength(2);
        expect(context.c.diagnostic).toMatchSnapshot();
        expect(context.c.data.application).toMatchSnapshot();
    });
});

describe('context reference', () => {
    test('simple', () => {
        const context = createSimpleContext();

        context.createReference({
            label: createSimpleTextNode('ref-1'),
        } as any as ContextReferenceContentInfo);
        context.createReference({
            label: createSimpleTextNode('ref-2'),
        } as any as ContextReferenceContentInfo);
        expect(
            context.getReferenceLabelIndex(createSimpleTextNode('ref-2')),
        ).toEqual(0);
        expect(
            context.getReferenceLabelIndex(createSimpleTextNode('ref-1')),
        ).toEqual(1);

        context.diagnoseAll();

        expect(context.c.diagnostic).toHaveLength(0);
        expect(context.c.data.reference).toMatchSnapshot();
    });

    test('create same key twice', () => {
        const context = createSimpleContext();

        context.createReference({
            label: createSimpleTextNode('ref-1'),
        } as any as ContextReferenceContentInfo);
        context.createReference({
            label: createSimpleTextNode('ref-1'),
        } as any as ContextReferenceContentInfo);

        context.diagnoseAll();

        expect(context.c.diagnostic).toHaveLength(2);
        expect(context.c.diagnostic).toMatchSnapshot();
        expect(context.c.data.reference).toMatchSnapshot();
    });

    test('create undefined keys', () => {
        const context = createSimpleContext();

        context.getReferenceLabelIndex(createSimpleTextNode('ref-1'));
        context.getReferenceLabelIndex(createSimpleTextNode('ref-2'));

        context.diagnoseAll();

        expect(context.c.diagnostic).toHaveLength(2);
        expect(context.c.diagnostic).toMatchSnapshot();
        expect(context.c.data.reference).toMatchSnapshot();
    });

    test('create unused keys', () => {
        const context = createSimpleContext();

        context.createReference({
            label: createSimpleTextNode('ref-1'),
        } as any as ContextReferenceContentInfo);
        context.createReference({
            label: createSimpleTextNode('ref-2'),
        } as any as ContextReferenceContentInfo);

        context.diagnoseAll();

        expect(context.c.diagnostic).toHaveLength(2);
        expect(context.c.diagnostic).toMatchSnapshot();
        expect(context.c.data.reference).toMatchSnapshot();
    });
});
