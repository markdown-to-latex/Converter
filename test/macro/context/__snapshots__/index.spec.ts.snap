// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`context application create same key twice 1`] = `
Array [
  Object {
    "errorType": 2,
    "filePath": "null",
    "message": "Application with app-1 already exist",
    "pos": Object {
      "end": Object {
        "absolute": 1,
        "column": 1,
        "line": 1,
      },
      "start": Object {
        "absolute": 0,
        "column": 1,
        "line": 1,
      },
    },
    "severity": "WARNING",
  },
  Object {
    "errorType": 2,
    "filePath": "null",
    "message": "Unused application label app-1",
    "pos": Object {
      "end": Object {
        "absolute": 1,
        "column": 1,
        "line": 1,
      },
      "start": Object {
        "absolute": 0,
        "column": 1,
        "line": 1,
      },
    },
    "severity": "WARNING",
  },
]
`;

exports[`context application create same key twice 2`] = `
Object {
  "labelToInfo": Object {
    "app-1": Object {
      "label": Object {
        "parent": null,
        "pos": Object {
          "end": 5,
          "start": 0,
        },
        "text": "app-1",
        "type": "Text",
      },
    },
  },
  "labelToRefs": Object {
    "app-1": Object {
      "node": Object {
        "parent": null,
        "pos": Object {
          "end": 1,
          "start": 0,
        },
        "type": "File",
      },
      "refs": 1,
    },
  },
  "labels": Array [],
}
`;

exports[`context application create undefined keys 1`] = `
Array [
  Object {
    "errorType": 2,
    "filePath": "null",
    "message": "Application with app-1 does not exist",
    "pos": Object {
      "end": Object {
        "absolute": 1,
        "column": 1,
        "line": 1,
      },
      "start": Object {
        "absolute": 0,
        "column": 1,
        "line": 1,
      },
    },
    "severity": "ERROR",
  },
  Object {
    "errorType": 2,
    "filePath": "null",
    "message": "Application with app-2 does not exist",
    "pos": Object {
      "end": Object {
        "absolute": 1,
        "column": 1,
        "line": 1,
      },
      "start": Object {
        "absolute": 0,
        "column": 1,
        "line": 1,
      },
    },
    "severity": "ERROR",
  },
]
`;

exports[`context application create undefined keys 2`] = `
Object {
  "labelToInfo": Object {},
  "labelToRefs": Object {},
  "labels": Array [],
}
`;

exports[`context application create unused keys 1`] = `
Array [
  Object {
    "errorType": 2,
    "filePath": "null",
    "message": "Unused application label app-1",
    "pos": Object {
      "end": Object {
        "absolute": 1,
        "column": 1,
        "line": 1,
      },
      "start": Object {
        "absolute": 0,
        "column": 1,
        "line": 1,
      },
    },
    "severity": "WARNING",
  },
  Object {
    "errorType": 2,
    "filePath": "null",
    "message": "Unused application label app-2",
    "pos": Object {
      "end": Object {
        "absolute": 1,
        "column": 1,
        "line": 1,
      },
      "start": Object {
        "absolute": 0,
        "column": 1,
        "line": 1,
      },
    },
    "severity": "WARNING",
  },
]
`;

exports[`context application create unused keys 2`] = `
Object {
  "labelToInfo": Object {
    "app-1": Object {
      "label": Object {
        "parent": null,
        "pos": Object {
          "end": 5,
          "start": 0,
        },
        "text": "app-1",
        "type": "Text",
      },
    },
    "app-2": Object {
      "label": Object {
        "parent": null,
        "pos": Object {
          "end": 5,
          "start": 0,
        },
        "text": "app-2",
        "type": "Text",
      },
    },
  },
  "labelToRefs": Object {
    "app-1": Object {
      "node": Object {
        "parent": null,
        "pos": Object {
          "end": 1,
          "start": 0,
        },
        "type": "File",
      },
      "refs": 1,
    },
    "app-2": Object {
      "node": Object {
        "parent": null,
        "pos": Object {
          "end": 1,
          "start": 0,
        },
        "type": "File",
      },
      "refs": 1,
    },
  },
  "labels": Array [],
}
`;

exports[`context application simple 1`] = `
Object {
  "labelToInfo": Object {
    "app-1": Object {
      "label": Object {
        "parent": null,
        "pos": Object {
          "end": 5,
          "start": 0,
        },
        "text": "app-1",
        "type": "Text",
      },
    },
    "app-2": Object {
      "label": Object {
        "parent": null,
        "pos": Object {
          "end": 5,
          "start": 0,
        },
        "text": "app-2",
        "type": "Text",
      },
    },
  },
  "labelToRefs": Object {
    "app-1": Object {
      "node": Object {
        "parent": null,
        "pos": Object {
          "end": 1,
          "start": 0,
        },
        "type": "File",
      },
      "refs": 2,
    },
    "app-2": Object {
      "node": Object {
        "parent": null,
        "pos": Object {
          "end": 1,
          "start": 0,
        },
        "type": "File",
      },
      "refs": 2,
    },
  },
  "labels": Array [
    "app-2",
    "app-1",
  ],
}
`;

exports[`context pictures create same key twice 1`] = `
Array [
  Object {
    "errorType": 2,
    "filePath": "null",
    "message": "Picture with label pic-1 already exists",
    "pos": Object {
      "end": Object {
        "absolute": 1,
        "column": 1,
        "line": 1,
      },
      "start": Object {
        "absolute": 0,
        "column": 1,
        "line": 1,
      },
    },
    "severity": "WARNING",
  },
]
`;

exports[`context pictures create same key twice 2`] = `
Object {
  "labelToInfo": Object {
    "pic-1": Object {
      "label": Object {
        "parent": null,
        "pos": Object {
          "end": 5,
          "start": 0,
        },
        "text": "pic-1",
        "type": "Text",
      },
    },
  },
  "labelToRefs": Object {
    "pic-1": Object {
      "node": Object {
        "parent": null,
        "pos": Object {
          "end": 1,
          "start": 0,
        },
        "type": "File",
      },
      "refs": 2,
    },
  },
  "labels": Array [
    "pic-1",
  ],
}
`;

exports[`context pictures create undefined keys 1`] = `
Array [
  Object {
    "errorType": 2,
    "filePath": "null",
    "message": "Undefined picture label pic-1",
    "pos": Object {
      "end": Object {
        "absolute": 1,
        "column": 1,
        "line": 1,
      },
      "start": Object {
        "absolute": 0,
        "column": 1,
        "line": 1,
      },
    },
    "severity": "ERROR",
  },
  Object {
    "errorType": 2,
    "filePath": "null",
    "message": "Undefined picture label pic-3",
    "pos": Object {
      "end": Object {
        "absolute": 1,
        "column": 1,
        "line": 1,
      },
      "start": Object {
        "absolute": 0,
        "column": 1,
        "line": 1,
      },
    },
    "severity": "ERROR",
  },
]
`;

exports[`context pictures create undefined keys 2`] = `
Object {
  "labelToInfo": Object {},
  "labelToRefs": Object {
    "pic-1": Object {
      "node": Object {
        "parent": null,
        "pos": Object {
          "end": 1,
          "start": 0,
        },
        "type": "File",
      },
      "refs": 1,
    },
    "pic-3": Object {
      "node": Object {
        "parent": null,
        "pos": Object {
          "end": 1,
          "start": 0,
        },
        "type": "File",
      },
      "refs": 1,
    },
  },
  "labels": Array [
    "pic-1",
    "pic-3",
  ],
}
`;

exports[`context pictures create unused keys 1`] = `
Array [
  Object {
    "errorType": 2,
    "filePath": "null",
    "message": "Unused picture label pic-1",
    "pos": Object {
      "end": Object {
        "absolute": 1,
        "column": 1,
        "line": 1,
      },
      "start": Object {
        "absolute": 0,
        "column": 1,
        "line": 1,
      },
    },
    "severity": "WARNING",
  },
  Object {
    "errorType": 2,
    "filePath": "null",
    "message": "Unused picture label pic-2",
    "pos": Object {
      "end": Object {
        "absolute": 1,
        "column": 1,
        "line": 1,
      },
      "start": Object {
        "absolute": 0,
        "column": 1,
        "line": 1,
      },
    },
    "severity": "WARNING",
  },
]
`;

exports[`context pictures create unused keys 2`] = `
Object {
  "labelToInfo": Object {
    "pic-1": Object {
      "label": Object {
        "parent": null,
        "pos": Object {
          "end": 5,
          "start": 0,
        },
        "text": "pic-1",
        "type": "Text",
      },
    },
    "pic-2": Object {
      "label": Object {
        "parent": null,
        "pos": Object {
          "end": 5,
          "start": 0,
        },
        "text": "pic-2",
        "type": "Text",
      },
    },
  },
  "labelToRefs": Object {
    "pic-1": Object {
      "node": Object {
        "parent": null,
        "pos": Object {
          "end": 1,
          "start": 0,
        },
        "type": "File",
      },
      "refs": 1,
    },
    "pic-2": Object {
      "node": Object {
        "parent": null,
        "pos": Object {
          "end": 1,
          "start": 0,
        },
        "type": "File",
      },
      "refs": 1,
    },
  },
  "labels": Array [
    "pic-1",
    "pic-2",
  ],
}
`;

exports[`context pictures simple 1`] = `
Object {
  "labelToInfo": Object {
    "pic-1": Object {
      "label": Object {
        "parent": null,
        "pos": Object {
          "end": 5,
          "start": 0,
        },
        "text": "pic-1",
        "type": "Text",
      },
    },
    "pic-2": Object {
      "label": Object {
        "parent": null,
        "pos": Object {
          "end": 5,
          "start": 0,
        },
        "text": "pic-2",
        "type": "Text",
      },
    },
    "pic-3": Object {
      "label": Object {
        "parent": null,
        "pos": Object {
          "end": 5,
          "start": 0,
        },
        "text": "pic-3",
        "type": "Text",
      },
    },
  },
  "labelToRefs": Object {
    "pic-1": Object {
      "node": Object {
        "parent": null,
        "pos": Object {
          "end": 1,
          "start": 0,
        },
        "type": "File",
      },
      "refs": 2,
    },
    "pic-2": Object {
      "node": Object {
        "parent": null,
        "pos": Object {
          "end": 1,
          "start": 0,
        },
        "type": "File",
      },
      "refs": 2,
    },
    "pic-3": Object {
      "node": Object {
        "parent": null,
        "pos": Object {
          "end": 1,
          "start": 0,
        },
        "type": "File",
      },
      "refs": 2,
    },
  },
  "labels": Array [
    "pic-3",
    "pic-1",
    "pic-2",
  ],
}
`;

exports[`context reference create same key twice 1`] = `
Array [
  Object {
    "errorType": 2,
    "filePath": "null",
    "message": "Reference with ref-1 already exist",
    "pos": Object {
      "end": Object {
        "absolute": 1,
        "column": 1,
        "line": 1,
      },
      "start": Object {
        "absolute": 0,
        "column": 1,
        "line": 1,
      },
    },
    "severity": "WARNING",
  },
  Object {
    "errorType": 2,
    "filePath": "null",
    "message": "Unused reference label ref-1",
    "pos": Object {
      "end": Object {
        "absolute": 1,
        "column": 1,
        "line": 1,
      },
      "start": Object {
        "absolute": 0,
        "column": 1,
        "line": 1,
      },
    },
    "severity": "WARNING",
  },
]
`;

exports[`context reference create same key twice 2`] = `
Object {
  "labelToInfo": Object {
    "ref-1": Object {
      "label": Object {
        "parent": null,
        "pos": Object {
          "end": 5,
          "start": 0,
        },
        "text": "ref-1",
        "type": "Text",
      },
    },
  },
  "labelToRefs": Object {
    "ref-1": Object {
      "node": Object {
        "parent": null,
        "pos": Object {
          "end": 1,
          "start": 0,
        },
        "type": "File",
      },
      "refs": 1,
    },
  },
  "labels": Array [],
}
`;

exports[`context reference create undefined keys 1`] = `
Array [
  Object {
    "errorType": 2,
    "filePath": "null",
    "message": "Reference with ref-1 does not exist",
    "pos": Object {
      "end": Object {
        "absolute": 1,
        "column": 1,
        "line": 1,
      },
      "start": Object {
        "absolute": 0,
        "column": 1,
        "line": 1,
      },
    },
    "severity": "ERROR",
  },
  Object {
    "errorType": 2,
    "filePath": "null",
    "message": "Reference with ref-2 does not exist",
    "pos": Object {
      "end": Object {
        "absolute": 1,
        "column": 1,
        "line": 1,
      },
      "start": Object {
        "absolute": 0,
        "column": 1,
        "line": 1,
      },
    },
    "severity": "ERROR",
  },
]
`;

exports[`context reference create undefined keys 2`] = `
Object {
  "labelToInfo": Object {},
  "labelToRefs": Object {},
  "labels": Array [],
}
`;

exports[`context reference create unused keys 1`] = `
Array [
  Object {
    "errorType": 2,
    "filePath": "null",
    "message": "Unused reference label ref-1",
    "pos": Object {
      "end": Object {
        "absolute": 1,
        "column": 1,
        "line": 1,
      },
      "start": Object {
        "absolute": 0,
        "column": 1,
        "line": 1,
      },
    },
    "severity": "WARNING",
  },
  Object {
    "errorType": 2,
    "filePath": "null",
    "message": "Unused reference label ref-2",
    "pos": Object {
      "end": Object {
        "absolute": 1,
        "column": 1,
        "line": 1,
      },
      "start": Object {
        "absolute": 0,
        "column": 1,
        "line": 1,
      },
    },
    "severity": "WARNING",
  },
]
`;

exports[`context reference create unused keys 2`] = `
Object {
  "labelToInfo": Object {
    "ref-1": Object {
      "label": Object {
        "parent": null,
        "pos": Object {
          "end": 5,
          "start": 0,
        },
        "text": "ref-1",
        "type": "Text",
      },
    },
    "ref-2": Object {
      "label": Object {
        "parent": null,
        "pos": Object {
          "end": 5,
          "start": 0,
        },
        "text": "ref-2",
        "type": "Text",
      },
    },
  },
  "labelToRefs": Object {
    "ref-1": Object {
      "node": Object {
        "parent": null,
        "pos": Object {
          "end": 1,
          "start": 0,
        },
        "type": "File",
      },
      "refs": 1,
    },
    "ref-2": Object {
      "node": Object {
        "parent": null,
        "pos": Object {
          "end": 1,
          "start": 0,
        },
        "type": "File",
      },
      "refs": 1,
    },
  },
  "labels": Array [],
}
`;

exports[`context reference simple 1`] = `
Object {
  "labelToInfo": Object {
    "ref-1": Object {
      "label": Object {
        "parent": null,
        "pos": Object {
          "end": 5,
          "start": 0,
        },
        "text": "ref-1",
        "type": "Text",
      },
    },
    "ref-2": Object {
      "label": Object {
        "parent": null,
        "pos": Object {
          "end": 5,
          "start": 0,
        },
        "text": "ref-2",
        "type": "Text",
      },
    },
  },
  "labelToRefs": Object {
    "ref-1": Object {
      "node": Object {
        "parent": null,
        "pos": Object {
          "end": 1,
          "start": 0,
        },
        "type": "File",
      },
      "refs": 2,
    },
    "ref-2": Object {
      "node": Object {
        "parent": null,
        "pos": Object {
          "end": 1,
          "start": 0,
        },
        "type": "File",
      },
      "refs": 2,
    },
  },
  "labels": Array [
    "ref-2",
    "ref-1",
  ],
}
`;

exports[`context tables create same key twice 1`] = `
Array [
  Object {
    "errorType": 2,
    "filePath": "null",
    "message": "Table with label [object Object] already exists",
    "pos": Object {
      "end": Object {
        "absolute": 1,
        "column": 1,
        "line": 1,
      },
      "start": Object {
        "absolute": 0,
        "column": 1,
        "line": 1,
      },
    },
    "severity": "WARNING",
  },
]
`;

exports[`context tables create same key twice 2`] = `
Object {
  "labelToInfo": Object {
    "table-1": Object {
      "label": Object {
        "parent": null,
        "pos": Object {
          "end": 7,
          "start": 0,
        },
        "text": "table-1",
        "type": "Text",
      },
    },
  },
  "labelToRefs": Object {
    "table-1": Object {
      "node": Object {
        "parent": null,
        "pos": Object {
          "end": 1,
          "start": 0,
        },
        "type": "File",
      },
      "refs": 2,
    },
  },
  "labels": Array [
    "table-1",
  ],
}
`;

exports[`context tables create undefined keys 1`] = `
Array [
  Object {
    "errorType": 2,
    "filePath": "null",
    "message": "Undefined table label table-1",
    "pos": Object {
      "end": Object {
        "absolute": 1,
        "column": 1,
        "line": 1,
      },
      "start": Object {
        "absolute": 0,
        "column": 1,
        "line": 1,
      },
    },
    "severity": "ERROR",
  },
  Object {
    "errorType": 2,
    "filePath": "null",
    "message": "Undefined table label table-3",
    "pos": Object {
      "end": Object {
        "absolute": 1,
        "column": 1,
        "line": 1,
      },
      "start": Object {
        "absolute": 0,
        "column": 1,
        "line": 1,
      },
    },
    "severity": "ERROR",
  },
]
`;

exports[`context tables create undefined keys 2`] = `
Object {
  "labelToInfo": Object {},
  "labelToRefs": Object {
    "table-1": Object {
      "node": Object {
        "parent": null,
        "pos": Object {
          "end": 1,
          "start": 0,
        },
        "type": "File",
      },
      "refs": 1,
    },
    "table-3": Object {
      "node": Object {
        "parent": null,
        "pos": Object {
          "end": 1,
          "start": 0,
        },
        "type": "File",
      },
      "refs": 1,
    },
  },
  "labels": Array [
    "table-1",
    "table-3",
  ],
}
`;

exports[`context tables create unused keys 1`] = `
Array [
  Object {
    "errorType": 2,
    "filePath": "null",
    "message": "Unused table label table-1",
    "pos": Object {
      "end": Object {
        "absolute": 1,
        "column": 1,
        "line": 1,
      },
      "start": Object {
        "absolute": 0,
        "column": 1,
        "line": 1,
      },
    },
    "severity": "WARNING",
  },
  Object {
    "errorType": 2,
    "filePath": "null",
    "message": "Unused table label table-2",
    "pos": Object {
      "end": Object {
        "absolute": 1,
        "column": 1,
        "line": 1,
      },
      "start": Object {
        "absolute": 0,
        "column": 1,
        "line": 1,
      },
    },
    "severity": "WARNING",
  },
]
`;

exports[`context tables create unused keys 2`] = `
Object {
  "labelToInfo": Object {
    "table-1": Object {
      "label": Object {
        "parent": null,
        "pos": Object {
          "end": 7,
          "start": 0,
        },
        "text": "table-1",
        "type": "Text",
      },
    },
    "table-2": Object {
      "label": Object {
        "parent": null,
        "pos": Object {
          "end": 7,
          "start": 0,
        },
        "text": "table-2",
        "type": "Text",
      },
    },
  },
  "labelToRefs": Object {
    "table-1": Object {
      "node": Object {
        "parent": null,
        "pos": Object {
          "end": 1,
          "start": 0,
        },
        "type": "File",
      },
      "refs": 1,
    },
    "table-2": Object {
      "node": Object {
        "parent": null,
        "pos": Object {
          "end": 1,
          "start": 0,
        },
        "type": "File",
      },
      "refs": 1,
    },
  },
  "labels": Array [
    "table-1",
    "table-2",
  ],
}
`;

exports[`context tables simple 1`] = `
Object {
  "labelToInfo": Object {
    "table-1": Object {
      "label": Object {
        "parent": null,
        "pos": Object {
          "end": 7,
          "start": 0,
        },
        "text": "table-1",
        "type": "Text",
      },
    },
    "table-2": Object {
      "label": Object {
        "parent": null,
        "pos": Object {
          "end": 7,
          "start": 0,
        },
        "text": "table-2",
        "type": "Text",
      },
    },
    "table-3": Object {
      "label": Object {
        "parent": null,
        "pos": Object {
          "end": 7,
          "start": 0,
        },
        "text": "table-3",
        "type": "Text",
      },
    },
  },
  "labelToRefs": Object {
    "table-1": Object {
      "node": Object {
        "parent": null,
        "pos": Object {
          "end": 1,
          "start": 0,
        },
        "type": "File",
      },
      "refs": 2,
    },
    "table-2": Object {
      "node": Object {
        "parent": null,
        "pos": Object {
          "end": 1,
          "start": 0,
        },
        "type": "File",
      },
      "refs": 2,
    },
    "table-3": Object {
      "node": Object {
        "parent": null,
        "pos": Object {
          "end": 1,
          "start": 0,
        },
        "type": "File",
      },
      "refs": 2,
    },
  },
  "labels": Array [
    "table-3",
    "table-1",
    "table-2",
  ],
}
`;
